Text/Distinct/Thesaurus.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft. All rights reserved.  
+// Licensed under the MIT License. See LICENSE file in the project root for full license information.  
+
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace AzureCognitiveSearch.PowerSkills.Text.Distinct
+{
+    public class Thesaurus
+    {
+        public Thesaurus(string executingDirectoryPath)
+            : this(JsonConvert.DeserializeObject<IEnumerable<IEnumerable<string>>>(
+                File.ReadAllText($"{executingDirectoryPath}\\thesaurus.json")))
+        { }
+
+        public Thesaurus(IEnumerable<IEnumerable<string>> dataset)
+        {
+            Synonyms = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
+            foreach (IEnumerable<string> lemma in dataset)
+            {
+                string canonicalForm = lemma.First();
+                foreach (string form in lemma)
+                {
+                    Synonyms.Add(Normalize(form), canonicalForm);
+                }
+            }
+        }
+
+        public Dictionary<string, string> Synonyms
+        {
+            get; private set;
+        }
+
+        public IEnumerable<string> Dedupe(IEnumerable<string> words)
+        {
+            var normalizedToWord = new Dictionary<string, string>();
+            foreach (string word in words)
+            {
+                string normalized = Normalize(word);
+                string canonical = Synonyms.TryGetValue(normalized, out string canonicalFromThesaurus) ?
+                    canonicalFromThesaurus :
+                    normalized;
+                if (!normalizedToWord.ContainsKey(canonical))
+                {
+                    normalizedToWord.Add(canonical, canonicalFromThesaurus ?? word); // Arbitrarily consider the first occurrence as canonical
```

Comment:
>word [](start = 78, length = 4)

do you want to use the normalized form here?

would be kind of weird if the distinct word that came back was like "time...." because the first occurrence was "Once upon a time..." even though that word doesn't appear as such anywhere else in the list


If you do decide to go with that, you can probably make normalizedToWord just a hash set

Reply:
I had the same concern, and actually decided against it because the normalized form is actually pretty weird when spaces and punctuation are involved. The reasoning here is that the normalized form is considered an internal representation used for comparisons, but should not constitute output. So sure, using the first occurrence is arbitrary and can be weird, but that can be mitigated by adding an entry with an explicit canonical form to the thesaurus. I actually don't think it will be a big issue in the context of extracted entities, which was the main scenario motivating building this skill.

And yes, I implemented it first with a hash set, and it was wayyyy simpler ;)